### ISSUE TITLE
Path conflict problem with Boot 2.4.x in Eclipse

### ISSUE THREAD

**Issue by evmcl** [spring-projects/sts4#574]
Trying to upgrade my project from Spring Boot 2.3.5 to 2.4.1. Using a Gradle based project that generates the eclipse `.classpath` and `.project` files using the eclipse-wtp plugin.

The project runs fine with `./gradlew bootRun`, however when I try to run from within Eclipse I get the following error:

```
2020-12-14 10:56:56.997 ERROR 321225 --- [  restartedMain] o.s.b.d.LoggingFailureAnalysisReporter   : 

***************************
APPLICATION FAILED TO START
***************************

Description:

An attempt was made to call a method that does not exist. The attempt was made from the following location:

    org.springframework.boot.SpringApplication.run(SpringApplication.java:324)

The following method did not exist:

    'void org.springframework.context.ConfigurableApplicationContext.setApplicationStartup(org.springframework.core.metrics.ApplicationStartup)'

The method's class, org.springframework.context.ConfigurableApplicationContext, is available from the following locations:

    jar:file:/home/user/.gradle/caches/modules-2/files-2.1/org.springframework/spring-context/5.2.9.RELEASE/4003ef2db8b5e4b22330fc6d67aae7ac5d304319/spring-context-5.2.9.RELEASE.jar!/org/springframework/context/ConfigurableApplicationContext.class
    jar:file:/home/user/.gradle/caches/modules-2/files-2.1/org.springframework/spring-context/5.3.2/83ff3909a582eceb5c8ae152f1e510672106f42b/spring-context-5.3.2.jar!/org/springframework/context/ConfigurableApplicationContext.class

The class hierarchy was loaded from the following locations:

    org.springframework.context.ConfigurableApplicationContext: file:/home/user/.gradle/caches/modules-2/files-2.1/org.springframework/spring-context/5.2.9.RELEASE/4003ef2db8b5e4b22330fc6d67aae7ac5d304319/spring-context-5.2.9.RELEASE.jar


Action:

Correct the classpath of your application so that it contains a single, compatible version of org.springframework.context.ConfigurableApplicationContext
```

The 5.3.2 version of `spring-context` is in the Eclipse `.classpath` file and I assume is the version I would want the project to use. I cannot figure out where the 5.2.9 version is coming from and why it is being injected by Eclipse/STS4 or how to remove it.

**Comment by evmcl** [comment]
FYI, just re-tested with the 4.9.0 release of STS4 and still seeing the same issue, although the version number of the first conflicting jar file has changed from 5.2.9 to 5.2.11.

**Comment by kdvolder** [comment]
We'd have to determine whether this is problem with STS, with Gradle or with Eclipse BuildShip. 

Can you provide details on how you run the project exactly. STS/Eclipse provides multiple ways to run projects. 

If I had to make a guess, I would assume the problem is mostly likely due to a bug in how BuildShip computes classpath for running the project. But it could also be a bug in STS launcher for Boot projects. 

A quick experiment you can try to distinguish between these two cases is:

- first run the project with "Run As >> Java Application"
- then run the project with "Run As >> Spring Boot App"

If both exhibit the same buggy behavior the bug is most likely in BuildShip and bug should be filed with them.

If only the "Run As >> Spring Boot App" has the problem then it means STS itself is probably to blame (i.e. BuildShip has the right classpath but STS somehow messes that up).

In that case we should try and fix it. But we'll need to be able to reproduce the issue so we can try to debug it first. For that it would be useful if you can provide a sample project that we can import and that demonstrates the poblem. This is crucial because the projects that we already tested with ourselves don't trigger this problem. That is not to say there is something wrong with your project of course. But there is something about your project the tools do not handle well, and we cannot begin to guess what it may be, without a means to reproduce the problem.

**Comment by martinlippert** [comment]
@evmcl Do you have any further details here with regards to the questions from @kdvolder ?

**Comment by evmcl** [comment]
After a lot of futzing about I figured out what was going on and have resolved it. In the end the problem was not with Eclipse/STS4.

I had a common library that was used by my Spring Boot app as well as other utilities. Within that common library the spring-security-core library was included as a dependency. Although I specified the same version (5.4.2) as used by the Spring Boot app, for whatever reasons it pulled in a different version of spring-context as a transitive dependency, which then got thrown into the .classpath file generated for Eclipse.

I added a transitive = false flag to the spring-security-core dependency in the common library to fix.
